### PR TITLE
pumbling: packfile, resolve external reference delta

### PR DESCRIPTION
**Pull request by ga-paul-t**
I don't really want to take any credit for this work. This solution is a carbon copy of what was proposed in https://github.com/go-git/go-git/pull/111

Would really appreciate some feedback

### CODE REVIEW

**Review by mcuadros — CHANGES_REQUESTED** [review]
can you provide a test?

**Comment by ga-paul-t** [comment]
@mcuadros I am working on adding a test this week. Just reading through and understanding the Git pack file specification. Apologies for the delay

**Comment by ga-paul-t** [comment]
@mcuadros I admit that I am struggling to write a test. I need to write a test where a packfile is a thin pack. Looking at the tests with regards to the packfile parser, they all pretty much use scenarios that clone existing Github repositories. As I have only encountered this issue with CodeCommit repositories, I can't follow that same pattern.

Would I need to construct a packfile manually? Or do you know if I can force this behaviour with a Github repository?

@bassofed wondering if you have any suggestions?

I really want this to be merged, as without it CodeCommit (or repositories that use thin packs) might not work

**Comment by bassofed** [comment]
> @mcuadros I admit that I am struggling to write a test. I need to write a test where a packfile is a thin pack. Looking at the tests with regards to the packfile parser, they all pretty much use scenarios that clone existing Github repositories. As I have only encountered this issue with CodeCommit repositories, I can't follow that same pattern.
> 
> Would I need to construct a packfile manually? Or do you know if I can force this behaviour with a Github repository?
> 
> @bassofed wondering if you have any suggestions?
> 
> I really want this to be merged, as without it CodeCommit (or repositories that use thin packs) might not work

Sorry for the delay.
I had the same problem. I saw this behavior in several CodeCommit repositories but they are private so I can't post in a example.

I faced the issue for the fist time in Portainer; if the pull fails, some functionality of Portainer are not working (https://github.com/portainer/portainer/issues/5185)

**Comment by ga-paul-t** [comment]
@mcuadros I have attempted to write my first test. I have tried to replicate this behaviour from a GitHub repository, but with no luck. As CodeCommit is a private service within AWS, there is no way to follow the same structure with the existing fixtures. As a middle ground, I have provided a packfile from a CodeCommit repository that triggers the issue with go-git without the proposed fix in this PR.

Interested in your feedback on this approach

**Comment by ga-paul-t** [comment]
@mcuadros any chance you could take a look. Thanks.

**Comment by mcuadros** [comment]
This approach is correct but you need to add the pack to the github.com/go-git/go-git-fixtures repository, where we host all the packfile fixtures, sorry about the delay. @ga-paul-t

**Comment by ga-paul-t** [comment]
@mcuadros I am attempting to add the packfile to the github.com/go-git/go-git-fixtures. I noticed there is a file called data.go, that is generated using "esc -o data.go -pkg=fixtures data".

I assume I need to regenerate this file?

**Comment by ga-paul-t** [comment]
@mcuadros I have raised a PR against `go-git-fixtures` repository as requested, https://github.com/go-git/go-git-fixtures/pull/4

Once merged. I will update the existing test to make use of this new fixture

**Comment by mcuadros** [comment]
@ga-paul-t merged the go-git-fixtures PR

**Comment by ga-paul-t** [comment]
> @ga-paul-t merged the go-git-fixtures PR

@mcuadros Could you tag the latest release in the go-git-fixtures repo. `go get github.com/go-git/go-git-fixtures/v4` doesn't pull down the test fixture change. Thanks.

**Comment by mcuadros** [comment]
My bad, I just tagged v4.3.0, @ga-paul-t

**Comment by ga-paul-t** [comment]
> My bad, I just tagged v4.3.0, @ga-paul-t

I'll update the PR later today

**Comment by ga-paul-t** [comment]
@mcuadros apologies but I just realised that I missed adding the .idx packfile to the `go-git-fixtures` repository. I have just raised another PR. Please accept my apologies for missing this. Link to the PR: https://github.com/go-git/go-git-fixtures/pull/5

Could you also retag when this has been merged? Thanks.

**Comment by mcuadros** [comment]
done

File: .gitignore [hunk 1]
```diff
@@ -2,3 +2,4 @@ coverage.out
 *~
 coverage.txt
 profile.out
+.tmp/
```
Comment (ga-paul-t) [inline .gitignore:5]: I added this to prevent accidental committing of temporary test files

**Comment by ga-paul-t** [comment]
@mcuadros that is now done. I have updated the PR to use the `go-git-fixtures` repo

**Comment by katehmd** [comment]
Is there any timeline for releasing a new go-git version that will include this PR?

**Comment by HDYA** [comment]
it seems cloning repo from Azure Devops still does not work, even after this is merged and go mod specifying it by git hash.
is this expected?
thanks

**Pull request merged**